torch/_dynamo/guards.py

```diff
@@ -413,23 +413,33 @@ def TENSOR_MATCH(self, guard: Guard):
             value = self.get(guard.name)
             assert isinstance(value, torch.Tensor)
             tensor_name = self.arg_ref(guard)
-            self.tensor_check_names.append(tensor_name)
-            self.tensor_check_examples.append(value)
-
-            # STOP - DO NOT USE id_ref FOR TENSORS - TENSOR INVALIDATION RULES DIFFER
-            self.tensor_check_ids[tensor_name] = id(value)
-
-            # Note: Guard code produced for tensor_match is a little different.
-            # We accumulate tensor names, then do a single install of `___check_tensors`.
-            # See _guards.cpp and TensorGuard for more information.
-            # TODO(voz): Add tensor matching code to export
-            # Note: this is a bit of a special case, and so does not use _produce_guard_code
-            guard.set_export_info(
-                "TENSOR_MATCH",
-                weakref.ref(type(value)),
-                None,
-                weakref.ref(value),
-            )
+            # [Note - On Export Tensor Guards]
+            #
+            # In eager mode, tensor guards are evaluated through C++, in guards.cpp
+            # see [Note - On Eager Tensor Guards] for more info.
+            #
+            # In export mode, we instead maintain parallel logic between C++ and python
+            # here, with an exception of checking the dispatch key - with the idea that a dispatch key
+            # is an entirely runtime notion that would make no sense to keep in an exported graph.
```

Comment:
This is not entirely true, although for Executorch it is sufficient. For example, suppose one of the input tensors had the negative dispatch key. You should end up with a graph that is specialized for tensors that have a negative dispatch key. If you allow a Tensor that does NOT have this bit set, you will accidentally run it "as if" it were negated.

Now, negative key only shows up for complex numbers, and furthermore Executorch doesn't support this feature at all, but the point stands that :some: tensor state only shows up on dispatch key. I think it's OK to omit guards on these, but we are obligated to error if a user passes in tensors with non-standard dispatch key